File: DSA/NarindVerma_500121898/Day12Q3.py
def compress_string(s: str) -> str:
    if not s:
        return s

    compressed = []
    count = 1

    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            count += 1
        else:
            compressed.append(s[i - 1] + str(count))
            count = 1
    compressed.append(s[-1] + str(count))

    compressed_str = ''.join(compressed)
    return compressed_str if len(compressed_str) < len(s) else s

# Test Cases
print(compress_string("aabcccccaaa"))  # a2b1c5a3
print(compress_string("abc"))         # abc
print(compress_string("aabbcc"))      # a2b2c2

File: DSA/Shlok_590017851/day12_q3.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
char* compressString(const char* str) 
{
    int len = strlen(str);
    char* compressed = (char*)malloc(2 * len + 1); 
    int count = 1, j = 0;
    for (int i = 1; i <= len; i++) 
    {
        if (str[i] == str[i - 1]) 
        {
            count++;
        } 
        else 
        {
            compressed[j++] = str[i - 1];
            j += sprintf(&compressed[j], "%d", count);
            count = 1;
        }
    }
    compressed[j] = '\0';
    if (strlen(compressed) >= len) {
        strcpy(compressed, str); 
    }
    return compressed;
}
int main()
{
    char* result1 = compressString("aabcccccaaa");
    char* result2 = compressString("abc");
    char* result3 = compressString("aabbcc");
  
    printf("%s\n", result1); 
    printf("%s\n", result2); 
    printf("%s\n", result3); 

    free(result1);
    free(result2);
    free(result3);
    return 0;
}
